jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/queue/README.md

```diff
@@ -0,0 +1,347 @@
+ActionNodeInstance(NodeInstanceImpl).triggerNodeInstances(Deque<NodeInstanceTriggerAction>)
+ActionNodeInstance(NodeInstanceImpl).triggerNodeInstance(NodeInstance, String)
+ActionNodeInstance(NodeInstanceImpl).addNodeInstanceTrigger(NodeInstance, NodeInstance, String)
+RuleFlowProcessInstance.addNodeInstanceTriggerAction(NodeInstance, NodeInstance, String)
+RuleFlowProcessInstance.addProcessInstanceAction(ProcessInstanceAction)
+```
+
+The above patterns of execution continue until the process instance has completed or otherwise hit a save point. 
+
+As you might suspect, there's a lot more logic in the process engine than simply triggering node instances -- and because all of that logic
+is built around a recursive execution model, it also needs to split into smaller sized actions.
+
+### Queue-based idioms 
+
+One of the idioms that now occurs frequently in the code is this: 
+
+```java
+  boolean queueBased = isQueueBased();
+  if( queueBased ) {
+    // if queue-based, add action to do X (after Y)
+    getProcessInstance().addAfterInternalTriggerAction(this); 
+  }
+  
+  // do Y
+  exceptionScopeInstance.handleException(faultName, processInstance.getFaultData()); 
+  
+  if( ! queueBased ) {
+    // if NOT queue-based, do X
+    afterInternalTrigger(); 
+  }
+```
+   
+In other words: 
+
+1. If queue-based, add an action to do *`x`* to the queue to be executed **after we do** *`y`*.
+2. Do *`y`*.
+3. if *not* queue-based, do *`x`*.
+                
+Furthermore, in order to do this, some methods needed to be broken up into multiple methods. 
+
+For examples, see `afterInternalTrigger()`, `afterEntryActions(NodeInstance, String)` and the
+`afterExitActions(String, boolean)` (among others) the `NodeInstance` implementation classes. 
+
+These methods are then called in various `ProcessInstanceAction` implementations. 
+
+## Comparing queue-based event history to recursive event history
+
+Let's take a simple process that contains the following nodes: 
+
+    start ( node "A" ) --> script task ( node "B" ) --> end ( node "C" )
+ 
+One simple way to see the difference between the recursive execution and the queue-based execution 
+is to look at the history of events for this process. 
+
+In the table below, we compare the two event histories. However, please first look at what the different abbreviations mean:
+
+- `bnl` - before node leave
+- `bnt` - before node trigger
+- `ant` - after node trigger
+- `anl` - after node leave
+- `bps` - before process (instance) start
+- `aps` - after process (instance) start
+- `apc` - before process (instance) complete
+- `apc` - after process (instance) complete
+  
+If we compare the event histories of the two execution modes for the above process, we see the following:
+
+| Recursive | Queue Based |
+| --------- | ----------- | 
+| `bps`     | `bps`       |
+| `bnt-A`   | `bnt-A`     |
+| `bnl-A`   | `bnl-A`     |
+| `bnt-B`   | **`ant-A`** |
+| `bnl-B`   | **`anl-A`** |
```

Comment:
After looking inside the code, in NodeInstanceImpl.trigger() there is fireBeforeNodeTriggered() called, then there is internalTrigger() called. Inside internalTrigger there is triggerCompleted called inside which following nodes are triggered and both fireBeforeNodeLeft() and fireAfterNodeLeft() are called. And only after that back again inside the NodeInstanceImpl.trigger() there is a call to fireAfterNodeTriggered(). So in case of Queue Based it should be `bnt-A` `bnl-A` **`anl-A`** and finally **`ant-A`** if I am not mistaken.